### PR TITLE
Internal Pagination: Adds FeedRange Adoption

### PR DESCRIPTION
**Pull request by bchong95**
# Internal Pagination: Adds FeedRange Adoption

## Description

This PR updates the Pagination library to use FeedRange instead of PartitionKeyRange or partitionkey range id. This simplifies the API from having ExecuteQuery overloads on partition key range id, epk range, and logical partition key. This is also needed so that single partition async enumerators can target a logical partition instead of an entire physical partition. The alternative was to wire through physical partition along with an optional logical partition, but we might as well take advantage of the unified FeedRange story.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj [hunk 1]
```diff
@@ -153,6 +153,10 @@
     <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="FeedRange\Continuations\" />
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Microsoft.Azure.Cosmos.csproj:157]: nit: remove

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs [hunk 2]
```diff
@@ -331,7 +332,14 @@ internal static class CosmosQueryExecutionContextFactory
             return ParallelCrossPartitionQueryPipelineStage.MonadicCreate(
                 documentContainer: documentContainer,
                 sqlQuerySpec: inputParameters.SqlQuerySpec,
-                targetRanges: targetRanges,
+                targetRanges: targetRanges
+                    .Select(range => new FeedRangeEpk(
+                        new Documents.Routing.Range<string>(
+                            min: range.MinInclusive, 
+                            max: range.MaxExclusive, 
+                            isMaxInclusive: true,
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:340]: Isn't this incorrent? Shouldn't it be:

isMinInclusive: true,
isMaxInclusive: false

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs [hunk 3]
```diff
@@ -377,7 +385,14 @@ internal static class CosmosQueryExecutionContextFactory
                 executionEnvironment: inputParameters.ExecutionEnvironment,
                 documentContainer: documentContainer,
                 sqlQuerySpec: inputParameters.SqlQuerySpec,
-                targetRanges: targetRanges,
+                targetRanges: targetRanges
+                    .Select(range => new FeedRangeEpk(
+                        new Documents.Routing.Range<string>(
+                            min: range.MinInclusive,
+                            max: range.MaxExclusive,
+                            isMaxInclusive: true,
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:393]: same here

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs [hunk 4]
```diff
@@ -19,5 +19,12 @@ namespace Microsoft.Azure.Cosmos
     abstract class FeedIteratorInternal : FeedIterator
     {
         public abstract CosmosElement GetCosmosElementContinuationToken();
+
+        public static bool IsRetriableException(CosmosException cosmosException)
+        {
+            return ((int)cosmosException.StatusCode == 429)
+                || (cosmosException.StatusCode == System.Net.HttpStatusCode.RequestTimeout)
+                || (cosmosException.StatusCode == System.Net.HttpStatusCode.ServiceUnavailable);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs:27]: nit: Are these best mapped through Retry Handlers? Why would we retry on a 503?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs [hunk 5]
```diff
@@ -55,115 +54,114 @@ internal sealed class InMemoryContainer : IMonadicDocumentContainer
...
             }
 
-
-            if (partitionKeyRange.Id == null)
+            if (feedRange is FeedRangeEpk feedRangeEpk)
             {
                 // look for overlapping epk ranges.
-                PartitionKeyHash? start = partitionKeyRange.MinInclusive == string.Empty ? (PartitionKeyHash?)null : PartitionKeyHash.Parse(partitionKeyRange.MinInclusive);
-                PartitionKeyHash? end = partitionKeyRange.MaxExclusive == string.Empty ? (PartitionKeyHash?)null : PartitionKeyHash.Parse(partitionKeyRange.MaxExclusive);
+                PartitionKeyHash? start = feedRangeEpk.Range.Min == string.Empty ? (PartitionKeyHash?)null : PartitionKeyHash.Parse(feedRangeEpk.Range.Min);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs:100]: Reading this PR I learned how to use PartitionKeyHash, thanks for those utilities!

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/StandByFeedContinuationTokenTests.cs [hunk 6]
```diff
@@ -221,92 +221,6 @@ public void ConstructorWithRangeGeneratesSingleQueue()
             Assert.AreEqual(max, deserialized[0].Range.Max);
         }
 
-        [TestMethod]
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/StandByFeedContinuationTokenTests.cs:224]: Why do we need to delete these tests?

Comment (bchong95) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/StandByFeedContinuationTokenTests.cs:224]: oh this leaked from the change feed PR (I did both in the same branch and created separate PRs) to keep the diff minimal and coherent for each.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs:340]: good catch. I think the only reason this doesnt matter is we dont end up reading these values in the inner layers of the code (we just assume the range is min inclusive and max exclusive)

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs:27]: With the new IAsyncEnumerable<TryCatch<Page>> pattern all requests are technically retriable. I just whitelisted some of them so that  `while(feedIterator.HasMoreResults)` doesn't run in an infinite loop for excpetions that are guarenteed to persist. In this case I see 503 as a retriable exception ... eventually the service will come back up. While bad request will never suddenly become a valid request.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs (hunk 5 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Pagination/InMemoryContainer.cs:100]: No problem :)

Yeah I am particularly pleased with how using PartitionKeyHash we can simulate splits in memory to avoid having to rely on the service fabric emulator or manually mocking out responses.

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs (hunk 4 above)
Comment (FabianMeiswinkel) [reply Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs:27]: I guess the assumption is that 410 is going to be retried downstream on the transport level already - and if it bubbles up due to max retries exceeded it is ok to not retry here any further?

**Review by FabianMeiswinkel — DISMISSED** [review]
Thanks Brandon - agree with Matias - this looks great, code i very well structured and the PkRangeHash tooling etc. to simulate split and merge will become very useful.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/CosmosQueryContextCore.cs [hunk 7]
```diff
@@ -58,31 +55,26 @@ internal CosmosDiagnosticsContext GetAndResetDiagnostics()
             lock (this.diagnosticLock)
             {
                 CosmosDiagnosticsContext current = this.diagnosticsContext;
-                this.diagnosticsContext = CosmosDiagnosticsContext.Create(this.queryRequestOptions);
+                this.diagnosticsContext = CosmosDiagnosticsContext.Create(new RequestOptions());
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/CosmosQueryContextCore.cs:58]: Why did you remove query request options? That has different headers options like getting the custom factory for the diagnostic context?

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs (hunk 4 above)
Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs:27]: 503 not only happens due to 410s, that is the assumption on Direct mode. But 503 can also be coming from a Gateway response, in which case, is not always a recoverable state based on CRIs.

In fact, we have a retry attempt cross-region on the ClientRetryPolicy (not on the same region). So if the 503 came up to this point, it means, the Retry policy we have does not apply (for example, account is in a single region). 503 are not equal to 410s all the time.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CosmosQueryExecutionContextFactory.cs]: Okay it did end up mattering for the emulator tests (my in memory container just doesnt look at these values)

File: Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Resource/FeedIterators/FeedIteratorInternal.cs:27]: To clarify all the different exception types will get bubbled up to the user. This method does not determine if the IAsynEnumerator should retry the exception for the user. It merely determines if we set the "HasMoreResults" flag to false for the user. This is only there to support the old FeedIterator model. In the IAsyncEnumerator model there is always more results unless .MoveNextAsync returns false.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/CosmosQueryContextCore.cs (hunk 7 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/v3Query/CosmosQueryContextCore.cs:58]: storing QueryRequestOptions as state is an anti pattern for these context classes. There should be one CosmosQueryContextCore per collection and users should pass in request level values as parameters to the method. This allows us to resuse the class more.

File: Microsoft.Azure.Cosmos/src/FeedRange/FeedRanges/FeedRangePartitionKey.cs [hunk 8]
```diff
@@ -26,12 +26,19 @@ public FeedRangePartitionKey(PartitionKey partitionKey)
             string containerRid,
             Documents.PartitionKeyDefinition partitionKeyDefinition)
         {
-            return Task.FromResult(
-new List<Documents.Routing.Range<string>>
-{
+            try
```
Comment (FabianMeiswinkel) [inline Microsoft.Azure.Cosmos/src/FeedRange/FeedRanges/FeedRangePartitionKey.cs:29]: NIT remove try/catch/throw

**Review by FabianMeiswinkel — DISMISSED** [review]

File: Microsoft.Azure.Cosmos/src/FeedRange/FeedRanges/FeedRangePartitionKey.cs (hunk 8 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/FeedRange/FeedRanges/FeedRangePartitionKey.cs:29]: thanks. I was debugging something and forgot to remove this.

**Review by sboshra — APPROVED** [review]
